File: 010/cpp/main.cpp
class Solution {
public:
    std::vector<string> parsePattern(string p) {
        std::vector<string> ss;
        std::string s;
        for (char &c: p) {
            if (s.empty()) {
                s = c;
                continue;
            }
            
            if (c != '*') {
                ss.push_back(s);
                s = c;
            } else {
                ss.push_back(s + '*');
                s = "";
            }
        }
        if (!s.empty()) {
            ss.push_back(s);
        }
        return ss;
    } 

    std::vector<string> simplify(std::vector<string> ps) {
        if (ps.empty()) {
            return ps;
        }
        std::vector<string> ps0;
        string p = ps[0];
        for (int i = 1; i < ps.size(); i++) {
            string p0 = ps[i];
            if (p0 == p && p != ".*" && p.size() > 1) {
                continue;
            }
            ps0.push_back(p);
            p = p0;
        }
        ps0.push_back(p);
        return ps0;
    }

    bool match(string s, string p) {
        if (p == ".*") {
            return true;
        }
        if (p.size() > 1) {
            char p0 = p[0];
            for (char &c: s) {
                if (c != p0) {
                    return false;
                }
            }
            return true;
        }
        if (p == ".") {
            return s.size() == 1;
        }
        return s == p;
    }

    bool isMatch(string s, std::vector<string> ps, int k) {
        if (k >= ps.size()) {
            return s.empty();
        }
        for (int i = 0; i < s.size() + 1; i++) {
            string s1 = s.substr(0, i);
            if (!match(s1, ps[k])) {
                if (i == 0) {
                    continue;
                }
                return false;
            }
            string s2 = s.substr(i, s.size() - i);
            if (isMatch(s2, ps, k + 1)) {
                return true;
            }
        }
        return false;
    }

    bool isMatch(string s, string p) {
        std::vector<string> ps = simplify(parsePattern(p));
        return isMatch(s, ps, 0);
    }
};

File: 010/rust/main.rs
impl Solution {
    pub fn parse(p: String) -> Vec<String> {
        let mut ps = Vec::new();
        let mut s = "".to_string();
        for c in p.chars() {
            if s == "" {
                s = c.to_string();
                continue;
            }
            if c == '*' {
                s.push(c);
                ps.push(s);
                s = "".to_string();
            } else {
                ps.push(s);
                s = c.to_string();
            }
        }
        if s != "" {
            ps.push(s);
        }
        return ps;
    }

    pub fn simplify(ps: Vec<String>) -> Vec<String> {
        let mut ps0 = Vec::new();
        if ps.is_empty() {
            return ps0;
        }
        let mut p = ps[0].clone();
        for i in 1..ps.len() {
            if ps[i] == p && p != ".*" && p.len() > 1 {
                continue;
            }
            ps0.push(p);
            p = ps[i].clone();
        }
        ps0.push(p);
        return ps0;
    }

    pub fn is_match0(s: String, p: String) -> bool {
        if p == ".*" {
            return true;
        } 
        if p.len() > 1 {
            let c = p.as_bytes()[0] as char;
            for c0 in s.chars() {
                if c0 != c {
                    return false;
                }
            }
            return true;
        }
        if p == "." {
            return s.len() == 1;
        }
        return s == p;
    }

    pub fn is_match1(s: String, ps: &Vec<String>, k: usize) -> bool {
        if (k >= ps.len()) {
            return s.is_empty();
        }
        for i in 0..s.len() + 1 {
            let s1 = s[0..i].to_string();
            if !Solution::is_match0(s1, ps[k].clone()) {
                if (i == 0) {
                    continue;
                }
                return false;
            }
            let s2 = s[i..s.len()].to_string();
            if Solution::is_match1(s2, ps, k + 1) {
                return true;
            }
        }
        return false;
    }

    pub fn is_match(s: String, p: String) -> bool {
        let ps = Solution::simplify(Solution::parse(p));
        return Solution::is_match1(s, &ps, 0);
    }
}
